Add transport utilities

aaron0browne opened the pull request:
To support a simpler service construction style. Mimics `echo` interfaces.

bruth commented:
I think this is going in the right direction. A couple of immediate changes is to put this package in `arcus/pkg/transport/http` to make it clear its HTTP-based in case we want to support other transports such as gRPC.

Second, I don't think `Context` should be an interface since it models data in addition to behavior. The getters and setters (e.g. `GetUserID`) is a code smell in Go for misuse of interfaces. A concrete `Context` type can still implement the behaviors, but data should just be fields.

I am concerned there is coupling between the transport and application layers. For example, the user id, access token, and role information is self-contained and transport-agnostic. It should modeled this way and likely in a separate package that defines application-level constructs (i.e. post-transport). This was the reason why I deliberately separated out [the http adaptor from service handler](https://github.com/arcus/eventual/blob/master/example/signtermsofuse_service.go#L13-L18) and no transport-level stuff leaked into the inner handler. The proposed change [does leak transport concerns](https://github.com/arcus/iam/pull/47/files#diff-e992c957e7a9112b67f4abd8f4ecae66R12) since this new context type provides those behaviors.

Reflecting a bit more, it could be that you simply want to wrap `echo.Context` to simplify the interface and bake in some more logic (like handling protobuf messages)? I think that is good, but it should still remain separate from app/service handlers. The transport adaptor may have a *transport* context and it builds a separate *handler* context for passing the data into it.

aaron0browne commented on transport/http/group.go:
```diff
@@ -0,0 +1,101 @@
+	g.middleware = append(g.middleware, m...)
+}
+
+func (g *group) Pre(m ...interface{}) error {
```
Yeah, so, this adds *transport-level* middleware, which will have a different signature depending on the transport, so in order to include it in the `transport` package interfaces, I used `interface{}`. But maybe that's a sign that it shouldn't be in that interface???

aaron0browne commented:
Addressed comments and added a `Translator` so that downstream modules can define a custom translation from http to transport handler (for the redirect edge cases we discussed).

bruth commented on transport/transport.go:
```diff
@@ -0,0 +1,45 @@
+package transport
```
I actually missed that these interfaces were defined in the transport package. I don't think we need or should abstract how a transport is defined right now, unless you had some other insight?

I am on the fence of whether having `Handler` defined in transport makes sense. The job of a transport is to handle serialization/deserialization off the wire and representation of the message per the protocol (HTTP is both the transport and protocol in this case in contrast to gRPC which is a protocol, but uses HTTP/2 for the transport), and then conversion into the target message format we are defining per the Handler signature.

So I think Handler should exist elsewhere but the transport reference it.

aaron0browne replied:
That makes sense to me. Maybe put `Handler` in `pkg/arcus`? Yeah, I figured that the interface might change if/when we tried to extend with other transports, but it wouldn't hurt to take a stab at it and use it downstream.

bruth replied:
Sure. I err on the side of trying things rather than being safe.. realizing that we may need to update a lot of code if something doesn't work out. If you are fine with that then proceed with your suggestion.

aaron0browne commented:
Ready for re-review @bruth!

bruth commented on arcus/context.go:
```diff
@@ -0,0 +1,36 @@
+)
+
+type Context struct {
+	*pb.Context
```
This should be embedded as a value, not a pointer, otherwise trying to access a field on an uninitialized value will panic. But I am on the fence with it being embedded in the first place vs. being a private field. Assuming there is a simple way to construct the context in some adaptor as well as produce a serialization-ready copy (i.e. the protobuf value) for adding to the event, keeping it private would be preferred.

aaron0browne replied:
Yes, my bad on the first point.

On the second, I guess we could add `AccessToken`, etc fields to the `Context` and a pair of marshal/unmarshal functions, right?

bruth replied:
Yes, not if the `Proto` method will be used.

bruth commented on arcus/internal/pb/context.proto:
```diff
@@ -0,0 +1,26 @@
+  string access_token = 4;
+
+  // Version of the code that produced this event.
+  string code_version = 5;
```
Thinking about this more, for this to be useful, it would need to consist of the service name and the commit SHA.

aaron0browne replied:
Should we make it two fields, then?

bruth replied:
Ha, sorry it was more a dangling statement... I mean it can a concatenated string, like `iam@ab37cac`, but discrete fields would be fine also.

bruth commented on transport/http/middleware.go:
```diff
@@ -0,0 +1,52 @@
+
+		ctx = c.Request().Context().(*arcus.Context)
+		c.Request().Header.Set("x-arcus-user", ctx.PrincipalId)
+		c.Request().Header.Set("x-arcus-role", ctx.PrincipalRole)
```
So where are these fields be set on the context? This middleware is a bit confusing because the use of `parseToken` implies this is the first time the token is being accessed, otherwise it should already be set on the context by the time this middleware is called. The assumption is that an upstream middleware does the parsing and the lookup for the access token in order to get the principle information.

aaron0browne replied:
Right, so I was trying to right a transport-specific middleware that knows how to translate the `arcus.Context` values for HTTP transport. So this would be the first time the token is accessed. A later `arcus.Middleware` layer that is transport-unaware would use the token to get the principal information and set it on the `Context`. This middleware then retrieves the principal info and places it in the transport-specific spots. Does that make sense? I'm not at all certain about this, just trying to explain what I was thinking.

bruth replied:
To me, this middleware would extract data from the headers and set the values on the context so `next` would have this info available. Setting the request header after calling next only makes sense if `next` is doing the work of looking up the user. It also assumes there is another handler in the chain that will leverage this updated request context.

The iam proxy is responsible parsing the token, doing the lookup and setting the `x-arcus-*` headers. Since this is a one-time specific implementation, I think this library should be geared towards all the other services of assuming the headers have been set (they should still deal with and check if the values are *actually* set or not).

aaron0browne replied:
> Setting the request header after calling next only makes sense if next is doing the work of looking up the user.

Yes, `next` will call the rest of the HTTP middleware stack, then the `Adaptor`, then the `iam` authentication middleware, which can be written as a transport-unaware `arcus.Middleware`, since the token is already on the `Context`, instead of still in the header.

> It also assumes there is another handler in the chain that will leverage this updated request context.

Yes, the other handler in the chain is the proxy middleware that will forward the request to the appropriate backend, with the proper headers now set. But maybe that doesn't make sense? Will a proxied request ever reach this code after `next(c)`?

---

All of this was just to enable writing a transport-unaware authentication middleware that would get the token from the `Context` and write the principal id and role to the `Context` without having to know about headers or whatever mechanism there was for transmitting that data on the chosen transport. But it sounds like maybe that doesn't make sense and we should just have a transport aware authentication middleware.

bruth commented on transport/http/translator.go:
```diff
@@ -0,0 +1,65 @@
+	"github.com/labstack/echo"
+)
+
+type Translator func(req proto.Message, next arcus.Handler) echo.HandlerFunc
```
Simply because I've used the term "adaptor" before to describe this, are you in favor of "translator" instead?

aaron0browne replied:
"Adaptor" is totally fine by me, I'll switch it!

aaron0browne commented:
Ready for re-review!

aaron0browne commented:
Remembered our conversation about properly parsing query parameters and found the code you were referring to in `kit` and copied it over here.

aaron0browne commented:
Added docstrings and exported the `http` structs so that downstream packages can cast to them and set the `Adaptor` and call `UseForHTTP`, which are not part of the `transport` interfaces.

bruth commented:
I have one final suggestion. I don't really like having a package named `arcus` since that isn't a meaningful name. My suggestion would be `service`, i.e. a `service.Handler` and `service.Context`. Thoughts?

bruth commented on transport/http/server.go:
```diff
@@ -0,0 +1,137 @@
+	return s.echo.Shutdown(ctx)
+}
+
+func (s *Server) add(method, path string, reqType proto.Message, h arcus.Handler, m ...arcus.Middleware) transport.Route {
```
`h` is not being used anywhere?

bruth commented:
Also can you add a few tests and will also serve to demonstrate the usage?

> exported the http structs so that downstream packages can cast to them and set the Adaptor and call UseForHTTP, which are not part of the transport interfaces.

I am a bit confused by this statement. So a service handler is passed to `Query()` or `Command()` which gets wrapped in the default adaptor for the transport. In some cases, we will need to use a custom adaptor to handle different edge cases, so we want the ability to do this, correct?

I actually think the transport interfaces is what is causing pain here since it (intentionally) reduces the surface area to some common interface, but that is only useful if that common interface would be used by something else and that is not clear here. One counter example is gRPC which doesn't have the notice of paths or groups, the endpoints are initialized by passing a struct that implements the gRPC generated service definition.

We can discuss more later, but my suggestion would be to remove the transport interfaces all together and keep the http transport types concrete. The signatures can be updated to support HTTP needs.

I think the caller should be responsible for wrapped the service handler with the adaptor, rather than the server, group, etc. doing that implicitly (which also removes the need to pass some global adaptor down the stack).

Hopefully this makes sense?

aaron0browne commented:
OK, removed the interfaces, renamed `arcus` to `service`, and added tests. I like your new avatar pic, btw!
